### PR TITLE
Add possibility to insert header between the actions #46

### PR DESCRIPTION
**Pull request by Pols95**
Now it's possible to define http-response-headers to add/set/remove (default add) for the actions (of any type):
```
action.1.director=director_name
action.1.enabled=true
action.1.id=action_id
action.1.type=cache
action.1.headers=h1,h2

/* custom headers */

header.id=h1
header.name=Strict-Transport-Security
header.mode= remove

header.id=h2
header.name=MyHeader
header.value=MyHeaderValue
header.mode=add | set
```

### CODE REVIEW

File: carapace-server/src/main/java/org/carapaceproxy/server/RequestHandler.java [hunk 1]
```diff
@@ -558,6 +558,12 @@ public void receivedFromRemote(HttpObject msg, EndpointConnection connection) {
                 cleanResponseForCachedData(httpMessage);
             }
         }
+
+        // Custom response Headers
+        if (msg instanceof HttpResponse && action != null && action.responseHeaders != null) {
+            ((HttpResponse) msg).headers().add(action.responseHeaders);
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/RequestHandler.java]: a question:
should we clone the HttpResponse object ?
are we serving it from the cache ?

File: carapace-server/src/main/java/org/carapaceproxy/server/config/ActionConfiguration.java [hunk 2]
```diff
@@ -63,9 +67,20 @@ public String getType() {
         return type;
     }
 
+    public List<String> getHeaders() {
+        return headers;
+    }
+
+    public ActionConfiguration addHeader(String id) {
+        if (!headers.contains(id)) {
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/config/ActionConfiguration.java]: shouldn't this be treated as a configuration error ?

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 3]
```diff
@@ -96,20 +99,39 @@ public void configure(ConfigurationStore properties) throws ConfigurationNotVali
...
-                LOG.info("configured action " + id + " type=" + action + " enabled:" + enabled);
-                if (enabled) {
-                    ActionConfiguration config = new ActionConfiguration(id, action, director, file, code);
-                    addAction(config);
+                ActionConfiguration config = new ActionConfiguration(id, action, director, file, code);
+                String _headers = properties.getProperty(prefix + "headers", "");
+                if (!_headers.isEmpty()) {
+                    String[] _headersids = _headers.split(",");
+                    for (String headerId : _headersids) {
+                        if (!this.headers.containsKey(headerId)) {
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: how do we deal with spaces ?
I see two options:
- "trim" the values
- "report an error" as you are currently doing

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 4]
```diff
@@ -343,6 +385,18 @@ public MapResult map(HttpRequest request, String userId, String sessionId, Backe
         }
     }
 
+    private HttpHeaders headersForAction(ActionConfiguration action) {
+        HttpHeaders _headers = new DefaultHttpHeaders();
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: are we always creating a new object ?
is it really needed ?
we are on the 'hot path' we should save as much resources as possible

File: carapace-server/src/main/java/org/carapaceproxy/server/config/ActionConfiguration.java (hunk 2 above)
Comment (Pols95) [reply carapace-server/src/main/java/org/carapaceproxy/server/config/ActionConfiguration.java]: I treaded it just like DirectorConfiguration#addBackend

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 5]
```diff
@@ -96,20 +99,48 @@ public void configure(ConfigurationStore properties) throws ConfigurationNotVali
...
+                            if (header != null) {
+                                config.addCustomHeader(header);
+                            } else {
+                                throw new ConfigurationNotValidException("while configuring action '" + id + "': header '" + headerId + "' does not exist");
+                            }
+                        }
+                    }
                 }
+                addAction(config);
+                LOG.info("configured action " + id + " type=" + action + " enabled:" + enabled);
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: headers ?

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 6]
```diff
@@ -209,6 +240,27 @@ public StandardEndpointMapper() {
...
+                _mode = HeaderMode.HEADER_MODE_SET;
+                break;
+            case "add":
+                _mode = HeaderMode.HEADER_MODE_ADD;
+                break;
+            case "remove":
+                _mode = HeaderMode.HEADER_MODE_REMOVE;
+                break;
+            default:
+                throw new ConfigurationNotValidException("invalid value of mode for header " + id);
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: invalid value of mode '+mode+'

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 7]
```diff
@@ -359,4 +423,33 @@ public String getForceBackendParameter() {
         return forceBackendParameter;
     }
 
+    public static enum HeaderMode {
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: final and private ?

if not "private" it is better to have a top level class

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 8]
```diff
@@ -359,4 +423,33 @@ public String getForceBackendParameter() {
...
+        HEADER_MODE_ADD, HEADER_MODE_SET, HEADER_MODE_REMOVE
+    }
+
+    public static final class CustomHeader {
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: top level class ?

**Pull request merged**